src/dtx/dtx_common.c

```diff
@@ -0,0 +1,632 @@
+
+				if (cont->sc_closing) {
+					d_list_del(&dbca->dbca_link);
+					dtx_flush_committable(cont_hdl);
+					ds_cont_hdl_put(cont_hdl);
+					D_FREE_PTR(dbca);
+					goto check;
+				}
+
+				if (!cont->sc_dtx_aggregating)
+					vos_dtx_stat(cont->sc_hdl, &stat);
+			}
+		}
+
+		if (!cont->sc_dtx_aggregating &&
+		    ((stat.dtx_committed_count > DTX_AGG_THRESHOLD_CNT) ||
+		     (stat.dtx_oldest_committed_time != 0 &&
+		      dtx_hlc_age2sec(stat.dtx_oldest_committed_time) >
+				DTX_AGG_THRESHOLD_AGE_UPPER))) {
+			ds_cont_get(cont);
+			cont->sc_dtx_aggregating = 1;
+			rc = dss_ult_create(dtx_aggregate, cont,
+				DSS_ULT_AGGREGATE, DSS_TGT_SELF, 0, NULL);
+			if (rc != 0) {
+				cont->sc_dtx_aggregating = 0;
+				ds_cont_put(cont);
+			}
+		}
+
+check:
+		rc = ABT_future_test(dmi->dmi_xstream->dx_shutdown, &state);
+		D_ASSERTF(rc == ABT_SUCCESS, "%d\n", rc);
+		if (state == ABT_TRUE) {
+			while (!d_list_empty(&dmi->dmi_dtx_batched_list)) {
+				dbca = d_list_entry(
+					dmi->dmi_dtx_batched_list.next,
+					struct dtx_batched_commit_args,
+					dbca_link);
+				d_list_del(&dbca->dbca_link);
+				ds_cont_hdl_put(dbca->dbca_cont_hdl);
+				D_FREE_PTR(dbca);
+			}
+
+			break;
+		}
+
+		ABT_thread_yield();
+	}
+}
+
+/**
+ * Prepare the DTX handle in DRAM.
+ *
+ * XXX: Currently, we only support to prepare the DTX against single DAOS
+ *	object and single dkey.
+ *
+ * \param dti		[IN]	The DTX identifier.
+ * \param oid		[IN]	The target object (shard) ID.
+ * \param coh		[IN]	Container open handle.
+ * \param epoch		[IN]	Epoch for the DTX.
+ * \param dkey_hash	[IN]	Hash of the dkey to be modified if applicable.
+ * \param conflict	[IN]	Hash of the dkey to be modified if applicable.
+ * \param dti_cos	[IN,OUT]The DTX array to be committed because of shared.
+ * \param dti_cos_count [IN,OUT]The @dti_cos array size.
+ * \param pm_ver	[IN]	Pool map version for the DTX.
+ * \param intent	[IN]	The intent of related modification.
+ * \param leader	[IN]	The target (to be modified) is leader or not.
+ * \param dth		[OUT]	Pointer to the DTX handle.
+ *
+ * \return			Zero on success, negative value if error.
+ */
+int
+dtx_begin(struct dtx_id *dti, daos_unit_oid_t *oid, daos_handle_t coh,
+	  daos_epoch_t epoch, uint64_t dkey_hash,
+	  struct dtx_conflict_entry *conflict, struct dtx_id *dti_cos,
+	  int dti_cos_count, uint32_t pm_ver, uint32_t intent, bool leader,
+	  struct dtx_handle **dthp)
+{
+	struct dtx_handle	*dth;
+
+	if (leader) {
+		/* XXX: For leader case, we need to find out the potential
+		 *	conflict DTXs in the CoS cache, and append them to
+		 *	the dispatched RPC to non-leaders. Then non-leader
+		 *	replicas can commit them before real modifications
+		 *	to avoid availability trouble.
+		 */
+		D_ASSERT(dti_cos == NULL);
+		D_ASSERT(dti_cos_count == 0);
+
+		dti_cos_count = vos_dtx_list_cos(coh, oid, dkey_hash,
+				intent == DAOS_INTENT_UPDATE ?
+				DCLT_PUNCH : DCLT_PUNCH | DCLT_UPDATE,
+				DTX_THRESHOLD_COUNT, &dti_cos);
```

Comment:
is it possible to have another incoming TX, which arrives later then this one, but tx_end before this one, and it has COS conflict with this one?

Reply:
If the two DTXs are both for update, then no conflict. If one is for update, another is for punch, then the latter one will not begin because of the conflict.